presto-main/src/main/java/com/facebook/presto/cost/HistogramCalculator.java

```diff
@@ -0,0 +1,167 @@
+            }
+
+            if (range.length() == 0.0) {
+                return totalDistinctValues.map(distinct -> 1.0 / distinct);
+            }
+
+            if (!isNaN(range.getDistinctValuesCount())) {
+                return totalDistinctValues.map(distinct -> min(1.0, range.getDistinctValuesCount() / distinct));
+            }
+
+            return Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+        }
+
+        // in the case the range is a single value, this can occur if the input
+        // filter range is a single value (low == high) OR in the case that the
+        // bounds of the filter or this histogram are infinite.
+        // in the case of infinite bounds, we should return an estimate that
+        // correlates to the overlapping distinct values.
+        if (lowPercentile.equals(highPercentile)) {
+            if (!useHeuristics) {
+                return Estimate.zero();
+            }
+            // If one of the bounds is unknown, but both percentiles are equal,
+            // it's likely that a heuristic value was returned
+            if (max.isUnknown() || min.isUnknown()) {
+                return totalDistinctValues.flatMap(distinct -> lowPercentile.map(lowPercent -> distinct * lowPercent));
+            }
+
+            return totalDistinctValues.map(distinct -> 1.0 / distinct);
+        }
+
+        // in the case that we return the entire range, the returned factor percent should be
+        // proportional to the number of distinct values in the range
+        if (lowPercentile.equals(Estimate.zero()) && highPercentile.equals(Estimate.of(1.0)) && min.isUnknown() && max.isUnknown()) {
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+
+            if (totalDistinctValues.equals(Estimate.zero())) {
+                return Estimate.of(1.0);
+            }
+            return totalDistinctValues.flatMap(totalDistinct -> {
+                if (DoubleMath.fuzzyEquals(totalDistinct, 0.0, 1E-6)) {
+                    return Estimate.unknown();
+                }
+                return Estimate.of(min(1.0, range.getDistinctValuesCount() / totalDistinct));
+            })
+            // in the case totalDistinct is NaN or 0
+            .or(() -> Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR));
+        }
+
+        return Optional.of(lowPercentile)
+                .filter(lowPercent -> !lowPercent.isUnknown())
+                .map(Estimate::getValue)
+                .map(lowPercent -> Optional.of(highPercentile)
+                        .filter(highPercent -> !highPercent.isUnknown())
+                        .map(Estimate::getValue)
+                        .map(highPercent -> highPercent - lowPercent)
+                        .map(Estimate::of)
+                        .orElseGet(() -> Estimate.of(1.0)))
```

Comment:
Should we return `Estimate.of(1.0 - lowPercent)` here?

Reply:
I took a closer look again since I think this code was written before adding `map` to the `Estimate` type. Both `lowPercentile` and `highPercentile` should be available at this time, so I think we should just return `highPercent - lowPercent` always, and in the case of any unknown, just leave it to unknown as it shouldn't occur. Any other value for unknowns would just be a heuristic anyways.

The new version would just be: 

```java
return lowPercentile.flatMap(lowPercent -> highPercentile.map(highPercent -> highPercent - lowPercent));
```

Reply:
Thanks for the explanation, it's more clearer.